src/main/java/seedu/address/model/Model.java

```diff
@@ -105,4 +107,35 @@
      */
     void setEventTaskFilter(Predicate<? super EventTask> predicate);
 
+    ////undo redo
+    /**
+     * saves a copy of the current TaskBook, and the Command that causes the change from the previous TaskBook to the current TaskBook.
+     * @param command
+     */
+    void recordState(Command command);
+
+    /**
+     * resets the TaskBook to the TaskBook before it was changed by the most recent undo.
+     * @return the Command that was redone
+     * @throws HeadAtBoundaryException
+     */
+    Command redo() throws HeadAtBoundaryException;
+
+    /**
+     * resets the TaskBook to it's previous state.
+     * @return the Command that was undone
+     * @throws HeadAtBoundaryException
+     */
+    Command undo() throws HeadAtBoundaryException;
+
+    /**
+     *@return true if TaskBook has changed
+     *Unlikely the commits list will be empty since recordState() is called when ModelManager is initialised. Thus there will always be at least one Commit in commits.
+     */
+    boolean hasUncommittedChanges();
+
+    String printRedoables();
```

Comment:
@potatowagon And remove the `undo-redo/trace` commands for this round, as well as these printing methods.